display_list/dl_op_spy.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/dl_op_spy.h"
+
+namespace flutter {
+
+bool DlOpSpy::DidDraw() {
+  return did_draw_;
+}
+
+void DlOpSpy::setColor(DlColor color) {
+  if (color.isTransparent()) {
+    dirty_ = false;
+  } else {
+    dirty_ = true;
+  }
+}
+void DlOpSpy::setColorSource(const DlColorSource* source) {
+  const DlColorColorSource* color_source = source->asColor();
+  if (color_source && color_source->color().isTransparent()) {
+    dirty_ = false;
+    return;
+  }
+  dirty_ = true;
+}
+void DlOpSpy::drawColor(DlColor color, DlBlendMode mode) {
+  if (did_draw_ || mode == DlBlendMode::kDst) {
+    return;
+  }
+  did_draw_ = !color.isTransparent() && mode != DlBlendMode::kClear;
+}
+void DlOpSpy::drawPaint() {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_;
+}
+void DlOpSpy::drawLine(const SkPoint& p0, const SkPoint& p1) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && p0 != p1;
+}
+void DlOpSpy::drawRect(const SkRect& rect) {
+  if (did_draw_) {
+    return;
+  }
+  did_draw_ = dirty_ && !rect.isEmpty();
```

Comment:
I'd have to check, but I think an empty rect will draw something in stroke mode just like a zero length line. Maybe there is a check somewhere that avoids drawing an empty rect even in stroke mode, but I'd have to search to make sure.

Reply:
What does it draw? A single pixel? Do you know if there is an existing test or document for this we can check?